proposals/0123-block-revenue-distribution.md

```diff
@@ -0,0 +1,150 @@
+
+Due to the lack of core protocol support for distributing block revenue to
+stakers, validators have developed their own solutions which are not enforced by
+the core protocol. For example, the Cogent validator diverts part of its fee
+revenue to NFT holders. But it's up the NFT holders to audit and hold Cogent
+accountable to a specific commission rate.
+
+Another alternative is Jito's mechanism for block "tips" (not fees, but the idea
+is similar). Jito's validator implementation includes a tip distribution program
+which it instructs validator operators to divert all of their tips to but cannot
+enforce perfect compliance. It's up to stakers and the Jito team to audit
+compliance by validator operators. This mechanism requires trusting a
+third-party (in this case Jito) to calculate reward distribution in an accurate
+and fair manner. It also relies on using a merkle tree to distribute fees to
+all stake accounts and the distributed fees are not automatically staked in
+recipient stake accounts.
+
+### Partitioned reward calculation
+
+The proposed design requires that all stake rewards pool accounts are updated
+before processing any transactions in the first block of an epoch. This adds
+more computation overhead at epoch boundaries but this overhead is directly
+proportional to the number of vote accounts which are similarly updated at the
+beginning of each epoch when calculating stake inflation rewards. If this
+calculation overhead becomes an issue, the reward bookkeeping in stake rewards
+pool accounts will need to be modified so that rewards for the new epoch do not
+interfere will reward calculations for the previous epoch.
+
+## New Terminology
+
+Stake Rewards Pool Account: An account that records how many rewards from block
+fees and tips earned by a particular validator will be distributed to stake
+delegators at the end of an epoch.
+
+## Detailed Design
+
+### Block Revenue Collection
+
+After all transactions are processed in a block for a given leader, rather than
+collecting all block revenue into the validator identity account, the protocol
+will check if the validator's vote account has specified commission rates and
+collector addresses in the new vote account version described in SIMD-XXXX.
+Block revenue is composed of two revenue sources: block fees and block tips,
+which are configured by validators independently.
+
+If the commission rate and collector account for a revenue source aren't set,
+all revenue will be collected into the validator's identity account as before.
+If the commission rate and collector account for a revenue source *are*
+specified, the commission amount MUST be calculated by first multiplying the
+rate by the amount of revenue and then using integer division to divide by one
+hundred. Then the commission amount should be deposited into the specified
+collector address ONLY if the collector account is system program owned and
+would be rent-exempt after the deposit.
```

Comment:
@michaelh-laine you previous wrote

> When rewards distribution occurs the commission is applied to the collector account funds (commission paid to vote account) and balance included with rewards paid into stake accounts

But I think it would be better for validators if they could receive block fee revenue immediately after each block in order to have a stream of income to pay for voting costs rather than needing to wait for epoch boundaries. With this approach, an important thing to consider is how to design commission rate updates because it would be a little strange to allow commission to change multiple times throughout the epoch. Maybe the vote program should only support setting the commission for the upcoming epoch so that the current epoch commission rates are always fixed from the beginning of an epoch.

Reply:
> Commission change should not be permitted after 50% of the epoch has elapsed the same as it is for staking and should apply for an entire epoch, agreed there.

Hoping to make commission related updates only apply for the following epoch FYI. Will be updating https://github.com/solana-foundation/solana-improvement-documents/pull/185 to reflect that design.

> Your proposal states commission should be paid to a system program-owned account, but I think a major use case here would be for larger and institutional validators to automatically divert those funds into a multisig for security and opsec purposes - this then naturally isn't a system-program owned account.

Note that since you can define a system-program owned account that *isn't* your node validator id, it doesn't need to have a private key anymore and so you could use a PDA which is controlled by a multisig or other onchain program.